staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go

```diff
@@ -62,25 +62,40 @@ type shouldRespondWithRetryAfterFunc func() (*retryAfterParams, bool)
 //   - 'Connection: close': tear down the TCP connection
 //
 // TODO: is there a way to merge WithWaitGroup and this filter?
-func WithRetryAfter(handler http.Handler, shutdownDelayDurationElapsedCh <-chan struct{}) http.Handler {
+func WithRetryAfter(handler http.Handler, when ShouldRespondWithRetryAfterFunc) http.Handler {
+	// NOTE: both WithRetryAfter and WithWaitGroup must use the same exact isRequestExemptFunc 'isRequestExemptFromRetryAfter,
+	// otherwise SafeWaitGroup might wait indefinitely and will prevent the server from shutting down gracefully.
+	return withRetryAfter(handler, isRequestExemptFromRetryAfter, when)
+}
+
+// NewShouldRespondWithRetryAfterFunc returns a ShouldRespondWithRetryAfterFunc
+func NewShouldRespondWithRetryAfterFunc(shutdownSendRetryAfter bool, shutdownCh <-chan struct{}) ShouldRespondWithRetryAfterFunc {
+	if !shutdownSendRetryAfter {
+		return func() (*retryAfterParams, bool) {
+			return nil, false
+		}
+	}
+
+	return newShutdownRetryAfterFunc(shutdownCh)
+}
```

Comment:
see comment in config.go. This func does not add value. The signal->bool conversion belongs into a local helper in config.go.